src/goto-symex/builtin_functions.cpp

```diff
@@ -1550,6 +1550,155 @@ void goto_symext::intrinsic_memset(
   dereference(ret_ref, dereferencet::READ);
   symex_assign(code_assign2tc(ret_ref, arg0), false, cur_state->guard);
 }
+void goto_symext::intrinsic_memcpy(
+  reachability_treet &art,
+  const code_function_call2t &func_call)
+{
+  assert(func_call.operands.size() == 3 && "Wrong memcpy signature");
+
+  const execution_statet &ex_state = art.get_cur_state();
+  if (ex_state.cur_state->guard.is_false())
+    return;
+
+  expr2tc dst = func_call.operands[0];
+  expr2tc src = func_call.operands[1];
+  expr2tc n = func_call.operands[2];
+
+  cur_state->rename(dst);
+  cur_state->rename(src);
+  cur_state->rename(n);
+
+  //if either operand is literally the constant 0 pointer, defer to our C memcpy
+  simplify(dst);
+  simplify(src);
+  if (
+    (is_constant_int2t(dst) && to_constant_int2t(dst).value.is_zero()) ||
+    (is_constant_int2t(src) && to_constant_int2t(src).value.is_zero()))
+  {
+    log_debug("memcpy", "NULL pointer operand, falling back");
+    bump_call(func_call, "c:@F@__memcpy_impl");
+    return;
+  }
+  if (is_nil_expr(dst) || is_nil_expr(src))
+  {
+    log_debug("memcpy", "Source or destination is NULL, falling back");
+    bump_call(func_call, "c:@F@__memcpy_impl");
+    return;
+  }
+
+  // Only handle constant-size copies for now
+  simplify(n);
+  if (!is_constant_int2t(n))
+  {
+    log_debug("memcpy", "Symbolic size not supported, falling back");
+    bump_call(func_call, "c:@F@__memcpy_impl");
+    return;
+  }
+
```

Comment:
Can you add a comment on which situation we handle here? For example, do we just execute the code below if both src and dst point to a valid object?